### ISSUE TITLE
Scaling up number of subscribers doesn't increase consumption

### ISSUE THREAD

**Issue by CatalinMustata** [googleapis/python-pubsub#778]
I have a number of identical clients (that's not using FlowControl yet) that subscribe to the same subscription, through streaming pull (I remember noticing this when looking at a memory leak in the client, however the GCP metrics show streaming pull at zero):

`future = subscriber.subscribe(SUBSCRIPTION, callback, use_legacy_flow_control=True)`

Normally, we're using just a few K8S pods to process the messages, but when we see an increase in the number of unack'ed messages, we scale up our deployment and add more subscribers.

The interesting part is that we have noticed this scaling of our deployment doesn't do much. CPU on each subscriber goes down, as well as the network I/O per pod (though overall network I/O increases some 30%) - making it look like the same workload is now split between more subscribers, instead of them processing it faster.

<img width="943" alt="Screenshot 2022-09-05 at 15 00 23" src="https://user-images.githubusercontent.com/5999012/188446376-3709a0bd-0c36-4758-bea6-83d5ef2c03a4.png">

I did some reading over this issue: https://github.com/googleapis/python-pubsub/issues/237 (though its premise is a bit different) and I understood the distinction between the two buffers at work, which _might_ explain the issue, but honestly I don't think that's it. 

What's even more interesting is that, if I'm reading the graphs right we're always behind a few hundred messages (publish to ack delta) and we're always ack-ing less than we're publishing (~10m/s vs ~20m/s) - this is strange because it would mean that the ACK delta should increase, yet it does not. It could be that the metric is vague, since another panel down the `Ack message count by delivery type` does average around 20m/s, dipping to 16 when the number of unack'ed started to rise.

Moreover, the number of unack'ed messages stays constant to almost 0, until ack rate drops to about 7m/s, when it climbs to almost 3k - this again makes no sense to me, as the ack vs publish difference (20-7 = avg 13) multipled by the time this was low (9 mins) would put the high watermark at about 7k messages (9 * 60 * 13).

<img width="937" alt="Screenshot 2022-09-05 at 15 00 32" src="https://user-images.githubusercontent.com/5999012/188446502-5556461d-ea48-481b-98d5-629d80bd5e12.png">

To summarize, I'd be grateful if someone could explain these behaviours (and maybe suggest how we can improve our system):
- why did the ack rate drop when CPU/Mem were constant
- why don't publish & ack rates match when unacked count is constant almost zero
- why didn't throughput increase significantly when scaling from 4 to 17 subscribers (on the contrary, each started doing less work)

**Comment by acocuzzo** [comment]
@CatalinMustata  You may want to take a look at: https://cloud.google.com/blog/products/data-analytics/testing-cloud-pubsub-clients-to-maximize-streaming-performance
Regarding the subscriber throughput, as you scale up your subscriber clients, Cloud Pub/Sub will load balance between them. You should receive the same total number of messages. 
In terms of your other questions:
- ack latencies could play a role in ack rate dropping
- The ack rate vs publish rate could be different (even with ~0 backlog) if the client is using filters
- If the backlog is ~0, increasing the number of clients won't increase the throughput.
However, I recommend creating a support case with the specific subscription.

**Comment by anguillanneuf** [comment]
@CatalinMustata A couple of pointers to perhaps help clarify some of the concepts involved.

- in the **Health** section of the metrics dashboards for a single subscription, you should see a chart named **Delivery latency health score**, and it should show you a **utilization** score. If this score is 1, you probably already have a sufficient number of streams open for receiving messages. Generally speaking, if this score is 0, you should probably consider opening more streams (creating more subscribers) to pull messages from the subscription. 
- is it possible that your subscription has a filter applied? if it has, then it explains why your publish rate is higher than your ack rate because filtered messages are not getting pulled and acked. 
- what happens when you don't use flow control is that your subscribers will always pull more messages than they can process and store them in the "buffer". if those messages stay in the "buffer" too long without getting processed, their ack deadlines will expire, causing the Pub/Sub server to resend them. Are you observing an increased expired ack deadlines count in the **Expired ack deadlines count** chart (also in the Health section of the metrics dashboard for a single subscription)? 

my suspicion is the machines you have may not be processing messages fast enough. but let's see if the above is making sense first.

**Comment by anguillanneuf** [comment]
Thanks @CatalinMustata for sharing these data. They are illuminating. 

**Utilization** looked great before scaling up, with a score of 1. Although it appeared to get rid of expired ack deadline, it probably wasn't going to help you improve your CPU utilization. 

To improve your CPU utilization, try these two things:
1. Turn on server-side flow control by setting `use_legacy_flow_control` to false (should also be the default). See the [Python documentation](https://googleapis.dev/python/pubsub/latest/pubsub/subscriber/api/client.html#google.cloud.pubsub_v1.subscriber.client.Client.subscribe). The legacy flow control was only done on the client-side. Experiment with client-side flow control. How many messages do you expect your provisioned machine to process in parallel at one time?
1. Increase the number of threads you use for processing messages. In Java, this can be set with the executorProvider, as in this [example](https://cloud.google.com/pubsub/docs/pull#java_4). This will need to be roughly translated into Python.

**Comment by CatalinMustata** [comment]
Thank you, @anguillanneuf! This kind of confirms my suspicions, though at first sight it doesn't seem like the number of worker threads is easily adjustable in Python.

We will experiment with different settings and report back.

**Comment by anguillanneuf** [comment]
Of course! If this load test framework is of any help: https://github.com/GoogleCloudPlatform/pubsub/tree/master/load-test-framework/python_src/clients

**Comment by CatalinMustata** [comment]
Thank you! Fortunately we've already written a small script that generates data similar to our normal load.

I do have one outstanding question though: what could have caused the increase of unacked messages? Most metrics looks relatively constant until unacked starts rising...

**Comment by anguillanneuf** [comment]
Yeah, I agree with you here. I can only guess that scaling up introduces cold starts problems. Subscribers need to re-establish connections and open new streams. (If that's indeed the case, we need to confirm by counting how many stream reconnects are happening - this is not a public metric now. And if it's really that bad, we need to look further for solutions for these cold stars problems. All in all you raise a good point. I will bring it up and cycle back.)

> - when we started scaling up the oldest unacked msg age was already 7.52 mins (seems like a lot to me)

**Comment by acocuzzo** [comment]
This issue should be discussed in a support case, as the amount of support we can provide for a specific subscription is limited in a public forum.
